### PR TITLE
[DebugInfo] Adding DIExpression into SIL

### PR DESCRIPTION
**Pull request by mshockwave**
This PR adds DIExpression infrastructure into SIL.

Previously, we can not use `debug_value` to mark the updated (SIL) SSA of a sub-field inside an aggregate-type source variable. For example:
```swift
struct MyStruct {
    var x : Int
}
%2 = alloc_stack $MyStruct, name "my_struct", loc "file.swift":3:4, scope 3
...
%5 = struct_element_addr %2 : $*MyStruct, #MyStruct.x, loc ...
```
In this snippet we are not able to generate a correct LLVM debug info for `%5` using `debug_value` or `debug_value_addr`. Because the current SIL debug infrastructure lacks the equivalent feature of LLVM's `!DIExpression`. More specifically, the `DW_OP_fragment` operator.

This is not a problem in `-Onone`, but an outstanding one in optimized build where optimizations -- SROA for example -- might break an aggregate-type object apart or even eliminate the original aggregate type. In which case the debug info quality for variable values will degrade.

This PR (when it's finished) essentially adds functionalities similar to `!DIExpression` at the SIL level. Take the code above as the example, we're now able to generate the following SIL:
```swift
struct MyStruct {
    var x : Int
}
%2 = alloc_stack $MyStruct, name "my_struct", loc "file.swift":3:4, scope 3
...
%5 = struct_element_addr %2 : $*MyStruct, #MyStruct.x, loc ...
debug_value_addr %5 : $*Int, var, (name "my_struct", loc "file.swift":3:4, scope 3), type $MyStruct, expr op_fragment:#MyStruct.x, loc "file.swift":10:11, scope 3
```
Here are the new parameters / features added to `debug_value` / `debug_value_addr`:
 - The debug variable can have (optional) SILLocation and/or SILDebugScope attached (the `loc` and `scope` directives _inside_ the parens enclosing `name`), which represents the source location where it was originally declared. If this extra info is absent, the instruction's source location will be used.
 - The `type` directive can be used to indicate the type where the debug variable was originally declared. If this info is absent, the type of the SSA value (i.e. `%5` in this case) will be used.
 - Finally, the `expr` directive is used to carry the DIExpression that basically has a one-to-one mapping to LLVM's `!DIExpression`. Currently we only have `op_fragment` and `op_deref` operator, corresponding to `DW_OP_LLVM_fragment` and `DW_OP_deref`, respectively.

The part that is slightly controversial is probably how we carry these extra information on in-memory representation for the affected SIL instructions. This change will make `SILDebugVariable`, the class that represents debug variable and used by all debug-info-carrying instructions (including the `alloc_*` family), more expensive. Luckily most of the time `SILDebugVariable` is generated on-demand (it is actually not _stored_ inside aforementioned SIL instructions) so hopefully it won't affect conventional SIL instruction transformations (like copying or moving SIL instructions).

I'm splitting this PR into two commits: The first commit add DIExpression support into SIL instructions classes (e.g. `DebugValueInst`) and changes regarding SILParser / SILPrinter; The second one add IRGen support.

### CODE REVIEW

**Comment by mshockwave** [comment]
I'll add test cases shortly

File: include/swift/SIL/SILDebugInfoExpression.h
Comment (adrian-prantl) [inline include/swift/SIL/SILDebugInfoExpression.h:52]: clang-format

File: include/swift/SIL/SILDebugInfoExpression.h
Comment (adrian-prantl) [inline include/swift/SIL/SILDebugInfoExpression.h:35]: Are we using a K_ prefix elsewhere? I would probably just drop it.

File: include/swift/SIL/SILDebugInfoExpression.h
Comment (adrian-prantl) [inline include/swift/SIL/SILDebugInfoExpression.h:26]: It would be great to add Doxygen comments to all new classes.

File: include/swift/SIL/SILInstruction.h
Comment (adrian-prantl) [inline include/swift/SIL/SILInstruction.h:4578]: `///`

Comment (adrian-prantl) [reply include/swift/SIL/SILInstruction.h:4578]: Maybe define an enum instead of commenting?

File: include/swift/SIL/SILInstruction.h
Comment (adrian-prantl) [inline include/swift/SIL/SILInstruction.h:4669]: and reuse the same enum here?

File: lib/IRGen/IRGenDebugInfo.cpp
Comment (adrian-prantl) [inline lib/IRGen/IRGenDebugInfo.cpp:2346]: In DIExpression, we define a custom iterator to decode the uint64_t array. Maybe we want to copy this?

File: lib/SIL/IR/SILDebugInfoExpression.cpp
Comment (adrian-prantl) [inline lib/SIL/IR/SILDebugInfoExpression.cpp:1]: The `-*- C++ -*-` marker only makes sense in a `.h` file, where it's ambiguous whether the language is C or C++ based on the file name.

File: lib/IRGen/IRGenDebugInfo.cpp
Comment (mshockwave) [reply lib/IRGen/IRGenDebugInfo.cpp:2346]: I'm not sure I understand you correctly: SILDebugInfoExpression doesn't store a list of uin64_t but some high-level objects like `Decl*`, so I don't think we can reuse llvm::DIExpression's iterator here.

File: include/swift/SIL/SILDebugInfoExpression.h
Comment (mshockwave) [reply include/swift/SIL/SILDebugInfoExpression.h]: Since there is also a member variable called "Operator" in this class, I'll rename `K_Operator` into `OperatorKind` to avoid ambiguous (so does `K_Decl`).

**Comment by mshockwave** [comment]
I'll squash all the commits into one right before it's ready to merge (or I can do it right now if you prefer)

**Comment by mshockwave** [comment]
@swift-ci  test

**Comment by mshockwave** [comment]
@swift-ci please test

File: lib/IRGen/IRGenDebugInfo.cpp
Comment (adrian-prantl) [reply lib/IRGen/IRGenDebugInfo.cpp:2346]: Would it make sense to also define a (new) iterator _similar_ to DIExpression's that makes it convenient to iterate through the list of SILDIExprOperand one operator at a time. So operator++ would skip over the entire Fragment, including its two parameters.

Comment (mshockwave) [reply lib/IRGen/IRGenDebugInfo.cpp:2346]: That is actually a good idea! Will add this feature

**Comment by mshockwave** [comment]
@swift-ci please test

**Comment by mshockwave** [comment]
Add documentations
@swift-ci please test

**Comment by adrian-prantl** [comment]
Thanks, this is looking great! It would be good to also have test cases for the Verifier assertions that are added by this patch.

**Comment by adrian-prantl** [comment]
@jckarter Do you have any opinion on the proposed SIL syntax extension?

**Comment by mshockwave** [comment]
> Thanks, this is looking great! It would be good to also have test cases for the Verifier assertions that are added by this patch.

Sorry I almost missed this comment.
And hopefully the windows build bot is back to work 😛 

@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]
LGTM.

**Pull request merged**
